File: src/main/java/com/ordana/spelunkery/items/PortalFluidBottleItem.java
package com.ordana.spelunkery.items;

import com.ordana.spelunkery.reg.GameRulez;
import com.ordana.spelunkery.utils.LevelHelper;
import net.minecraft.advancements.CriteriaTriggers;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.stats.Stats;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.HoneyBottleItem;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.ItemUtils;
import net.minecraft.world.item.Items;
import net.minecraft.world.level.Level;
import org.jetbrains.annotations.NotNull;

import static net.minecraft.core.component.DataComponents.LODESTONE_TRACKER;

public class PortalFluidBottleItem extends HoneyBottleItem
{
	public PortalFluidBottleItem (Properties properties)
	{
		super(properties);
	}
	
	//Override
	public boolean shouldCauseReequipAnimation (ItemStack oldStack, ItemStack newStack, boolean slotChanged)
	{
		return false;
	}
	
	@Override
	public boolean isFoil (@NotNull ItemStack stack)
	{
		return true;
	}
	
	@Override
	@NotNull
	public SoundEvent getDrinkingSound ()
	{
		return SoundEvents.HONEY_DRINK;
	}
	
	@Override
	@NotNull
	public SoundEvent getEatingSound ()
	{
		return SoundEvents.HONEY_DRINK;
	}
	
	
	@Override
	@NotNull
	public ItemStack finishUsingItem (@NotNull ItemStack stack, @NotNull Level level, @NotNull LivingEntity livingEntity)
	{
		if (livingEntity instanceof Player player)
		{
			player.setItemInHand(
				player.getUsedItemHand(),
				ItemUtils.createFilledResult(stack, player, Items.GLASS_BOTTLE.getDefaultInstance())
			);
			
		}
		if (livingEntity instanceof ServerPlayer serverPlayer && level.getGameRules().getBoolean(GameRulez.BADLANDS_CHUGS_PORTALS))
		{
			CriteriaTriggers.CONSUME_ITEM.trigger(serverPlayer, stack);
			serverPlayer.awardStat(Stats.ITEM_USED.get(this));
			
			var uhh = stack.get(LODESTONE_TRACKER);
			if (uhh == null)
			{
				LevelHelper.teleportToSpawnPosition(serverPlayer);
			}
			else
			{
				uhh.target().ifPresent(globalPos -> LevelHelper.teleportToAnchorPosition(serverPlayer, globalPos));
			}
		}
		return stack;
	}
	
}

File: src/main/java/com/ordana/spelunkery/items/PortalFluidBottleItem.kt
package com.ordana.spelunkery.items

import com.ordana.spelunkery.reg.GameRulez
import com.ordana.spelunkery.utils.LevelHelper
import net.minecraft.ChatFormatting
import net.minecraft.advancements.CriteriaTriggers
import net.minecraft.core.component.DataComponents.LODESTONE_TRACKER
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerPlayer
import net.minecraft.sounds.SoundEvents
import net.minecraft.stats.Stats
import net.minecraft.world.effect.MobEffectInstance
import net.minecraft.world.effect.MobEffects
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.HoneyBottleItem
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.minecraft.world.item.TooltipFlag
import net.minecraft.world.level.Level

class PortalFluidBottleItem(properties: Properties) : HoneyBottleItem(properties)
{
	override fun shouldCauseReequipAnimation(oldStack: ItemStack, newStack: ItemStack, slotChanged: Boolean)
		= false

	override fun isFoil(stack: ItemStack) = true

	override fun getDrinkingSound() = SoundEvents.HONEY_DRINK

	override fun getEatingSound() = SoundEvents.HONEY_DRINK

	override fun appendHoverText(
		stack: ItemStack,
		context: TooltipContext,
		tooltipComponents: MutableList<Component?>,
		tooltipFlag: TooltipFlag
	)
	{
		stack.get(LODESTONE_TRACKER)?.target?.ifPresent {
			val co = it.pos
			val dim = it.dimension
			val cal = Component.translatable(
				"tooltip.spelunkery.portal_fluid_target",
				-co.z,
				+co.x,
				+co.y,
				Component.translatable(dim.location().toLanguageKey("dimension"))
			)

			cal.withStyle(ChatFormatting.GRAY)
			tooltipComponents += cal
		}

		super.appendHoverText(stack, context, tooltipComponents, tooltipFlag)
	}

	override fun finishUsingItem(stack: ItemStack, level: Level, user: LivingEntity): ItemStack
	{
		val eatRes = when (val fp = stack.getFoodProperties(user))
		{
			null -> stack
			else -> user.eat(level, stack, fp)
		}

		if (eatRes.isEmpty)
		{
			return ItemStack(Items.GLASS_BOTTLE)
		}
		else
		{
			if (user is Player)
			{
				user.addEffect(MobEffectInstance(MobEffects.CONFUSION, 4 * level.tickRateManager().tickrate().toInt(), 6))
				if (!user.hasInfiniteMaterials())
				{
					val itemstack = ItemStack(Items.GLASS_BOTTLE)
					if (!user.getInventory().add(itemstack))
					{
						user.drop(itemstack, false)
					}
				}

				if (level.gameRules.getBoolean(GameRulez.BADLANDS_CHUGS_PORTALS))
				{
					if (user is ServerPlayer)
					{
						CriteriaTriggers.CONSUME_ITEM.trigger(user, stack)
						user.awardStat(Stats.ITEM_USED.get(this))

						val uhh = stack.get(LODESTONE_TRACKER)
						if (uhh == null)
						{
							LevelHelper.teleportToSpawnPosition(user)
						}
						else
						{
							uhh.target.ifPresent { LevelHelper.teleportToAnchorPosition(user, it) }
						}
					}
				}
			}
			return eatRes
		}
	}
}
